### ISSUE TITLE
Abstract Class design pattern in Go

### ISSUE THREAD

**Issue by huan** [wechaty/go-wechaty#11]
Questions from WIP: add user/message.go #4

### Q3

> Q3: In my understanding, Puppet should be an interface definition, different protocols to implement different classes, but some methods of Puppet have logic, and they cannot have logic in Go's interface definition. I don't know what to do.  
>   
> Sorry, maybe my understanding of wechaty and wechat-puppet is too simple, and I may need to read and understand them as a whole.

If I'm understanding correctly, our problem is the Go Lang does not permit an abstract class to have any implementation in it?

Please feel free to correct me if I'm wrong.

To be updated... I'm going to study Go now...

**Comment by dchaofei** [comment]
Can write logic in "Puppet struct" method, I think #4  code is pretty good now

**Comment by suntong** [comment]
> Go Lang does not permit an abstract class to have any implementation in it

Not true, I've defined some common functions in my base / abstract class, taking a look at https://github.com/wechaty/go-wechaty/pull/4 now...

**Comment by suntong** [comment]
BTW, 

You know there is a tool to "Convert TypeScript **types** to GO" right? -- https://app.quicktype.io/#l=Go

**Comment by huan** [comment]
@suntong Thank you very much for letting us know the "Convert TypeScript types to GO" project!

I believe that's definitely we should use because the wechaty-puppet will be upgraded frequently and we have to build an automation process to sync other languages like Go from the TypeScript without needing any human helps.
